uast/npath.go

```diff
@@ -0,0 +1,225 @@
+package uast
+
+import (
+	"errors"
+)
+
+//CodeReference
+//https://pmd.github.io/pmd-5.7.0/pmd-java/xref/net/sourceforge/pmd/lang/java/rule/codesize/NPathComplexityRule.html
+
+//NpathComplexity return a slice of int for each function in the tree
+func NpathComplexity(n *Node) ([]int, error) {
+	var funcs []*Node
+	var npath []int
+
+	if n.containsRol(FunctionDeclarationBody) {
+		funcs = append(funcs, n)
+	} else {
+		funcs = n.deepChildrenOfRole(FunctionDeclarationBody)
+	}
+	if len(funcs) <= 0 {
+		npath = append(npath, -1)
+		return npath, errors.New("Function declaration not found")
+	}
+	for _, function := range funcs {
+		npath = append(npath, visitFunctionBody(function))
+	}
+	return npath, nil
+}
+
+func visitorSelector(n *Node) int {
+	// I need to add a error when the node dont have any rol
+	// when I got 2 or more roles in a node this doesnt work
+	for _, rol := range n.Roles {
+		switch rol {
+		case If:
+			return visitIf(n)
+		case While:
+			return visitWhile(n)
+		case Switch:
+			return visitSwitch(n)
+		case DoWhile:
+			return visitDoWhile(n)
+		case For:
+			return visitFor(n)
+		case Return:
+			return visitReturn(n)
+		default:
+			continue
+		}
+	}
+	return visitNotCompNode(n)
+}
+
+func complexityMultOf(n *Node) int {
+	npath := 1
+	for _, child := range n.Children {
+		npath *= visitorSelector(child)
+	}
+	return npath
+}
+
+func complexitySumOf(n *Node) int {
+	npath := 0
+	for _, child := range n.Children {
+		npath += visitorSelector(child)
+	}
+	return npath
+}
+
+func visitFunctionBody(n *Node) int {
+	return complexityMultOf(n)
+}
+
+func visitNotCompNode(n *Node) int {
+	return complexityMultOf(n)
+}
+
+func visitIf(n *Node) int {
+	// (npath of if + npath of else (or 1) + bool_comp of if) * npath of next
+	npath := 0
+	ifBody := n.childrenOfRole(IfBody)
+	ifCondition := n.childrenOfRole(IfCondition)
+	ifElse := n.childrenOfRole(IfElse)
+
+	if len(ifElse) == 0 {
+		npath++
+	} else {
+		npath += complexityMultOf(ifElse[0])
+	}
+	npath *= complexityMultOf(ifBody[0])
```

Comment:
shouldn't npath be added here, instead of multiplied? From the `npath of if + npath of else (or 1)` clause.
I see the same pattern in the following visitor functions, so maybe I'm missing something?